core/src/main/java/nu/yona/server/device/service/DeviceService.java

```diff
@@ -243,4 +255,94 @@ public void updateOperatingSystem(UUID userId, UUID deviceId, OperatingSystem op
 			deviceAnonymizedRepository.save(deviceAnonymized);
 		});
 	}
+
+	@Transactional
+	public void postOpenAppEvent(UUID userId, UUID deviceId, Optional<OperatingSystem> operatingSystem,
+			Optional<String> appVersion)
+	{
+		userService.updateUser(userId, userEntity -> {
+			operatingSystem.ifPresent(os -> assertValidAppVersion(os, appVersion.get()));
+
+			LocalDate now = TimeUtil.utcNow().toLocalDate();
+			UserDevice deviceEntity = getDeviceEntity(deviceId);
+			DeviceAnonymized deviceAnonymizedEntity = deviceEntity.getDeviceAnonymized();
+
+			setAppLastOpenedDateIfNewer(now, userEntity::getAppLastOpenedDate, () -> userEntity.setAppLastOpenedDate(now));
+			setAppLastOpenedDateIfNewer(now, () -> Optional.of(deviceEntity.getAppLastOpenedDate()),
+					() -> deviceEntity.setAppLastOpenedDate(now));
+			operatingSystem.ifPresent(os -> setOperatingSystemIfWasUnknown(deviceEntity, os));
+			appVersion.ifPresent(av -> setAppVersionIfDifferent(deviceAnonymizedEntity, av));
+		});
+	}
+
+	private void setAppLastOpenedDateIfNewer(LocalDate now, Supplier<Optional<LocalDate>> currentDateSupplier,
+			Runnable dateSetter)
+	{
+		Optional<LocalDate> appLastOpenedDate = currentDateSupplier.get();
+		if (!appLastOpenedDate.isPresent() || appLastOpenedDate.get().isBefore(now))
+		{
+			dateSetter.run();
+		}
+	}
+
+	private void setOperatingSystemIfWasUnknown(UserDevice deviceEntity, OperatingSystem currentOperatingSystem)
+	{
+		DeviceAnonymized deviceAnonymized = deviceEntity.getDeviceAnonymized();
+		OperatingSystem registeredOperatingSystem = deviceAnonymized.getOperatingSystem();
+		if (registeredOperatingSystem != currentOperatingSystem && registeredOperatingSystem != OperatingSystem.UNKNOWN)
+		{
+			throw DeviceServiceException.cannotSwitchDeviceOperatingSystem(registeredOperatingSystem, currentOperatingSystem,
+					deviceEntity.getId());
+		}
+		deviceAnonymized.setOperatingSystem(currentOperatingSystem);
+	}
+
+	private void setAppVersionIfDifferent(DeviceAnonymized deviceAnonymizedEntity, String appVersion)
+	{
+		if (!Objects.equals(deviceAnonymizedEntity.getAppVersion(), appVersion))
+		{
+			deviceAnonymizedEntity.setAppVersion(appVersion);
+		}
+	}
+
+	private void assertValidAppVersion(OperatingSystem operatingSystem, String appVersionStr)
+	{
+		String minAppVersion;
+		switch (operatingSystem)
+		{
+			case ANDROID:
+				minAppVersion = ANDROID_MIN_APP_VERSION;
+				break;
+			case IOS:
+				minAppVersion = IOS_MIN_APP_VERSION;
+				break;
+			case UNKNOWN:
+				// Everything is OK for as long as the app didn't confirm the operating system
+				return;
+			default:
+				throw new IllegalArgumentException("Unknown operating system: " + operatingSystem);
+		}
+		assertNotOlderThan(operatingSystem, minAppVersion, parseVersionString(appVersionStr));
+	}
+
+	private void assertNotOlderThan(OperatingSystem operatingSystem, String minAppVersionStr, Version appVersion)
```

Comment:
Please switch `operatingSystem` to the last parameter position.

Reply:
Done